pkg/mcs/tso/server/server.go

```diff
@@ -98,16 +130,170 @@ func (s *Server) Context() context.Context {
 
 // Run runs the TSO server.
 func (s *Server) Run() error {
+	go systimemon.StartMonitor(s.ctx, time.Now, func() {
+		log.Error("system time jumps backward", errs.ZapError(errs.ErrIncorrectSystemTime))
+		timeJumpBackCounter.Inc()
+	})
+
+	if err := s.initClient(); err != nil {
+		return err
+	}
+	if err := s.startServer(); err != nil {
+		return err
+	}
+
+	s.startServerLoop()
+
 	return nil
 }
 
+func (s *Server) startServerLoop() {
+	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(s.ctx)
+	s.serverLoopWg.Add(2)
+	go s.primaryElectionLoop()
+	go s.tsoAllocatorLoop()
+}
+
+// tsoAllocatorLoop is used to run the TSO Allocator updating daemon.
+func (s *Server) tsoAllocatorLoop() {
+	defer logutil.LogPanic()
+	defer s.serverLoopWg.Done()
+
+	ctx, cancel := context.WithCancel(s.serverLoopCtx)
+	defer cancel()
+	s.tsoAllocatorManager.AllocatorDaemon(ctx)
+	log.Info("tso server is closed, exit allocator loop")
+}
+
+func (s *Server) primaryElectionLoop() {
+	defer logutil.LogPanic()
+	defer s.serverLoopWg.Done()
+
+	for {
+		if s.IsClosed() {
+			log.Info("server is closed, exit tso primary election loop")
+			return
+		}
+
+		primary, rev, checkAgain := s.participant.CheckLeader()
+		if checkAgain {
+			continue
+		}
+		if primary != nil {
+			// TODO: if enable-local-tso is true, check the cluster dc-location after the primary/leader is elected
+			// go s.tsoAllocatorManager.ClusterDCLocationChecker()
+
+			log.Info("start to watch the primary/leader", zap.Stringer("tso-primary", primary))
+			// WatchLeader will keep looping and never return unless the primary/leader has changed.
+			s.participant.WatchLeader(s.serverLoopCtx, primary, rev)
+			log.Info("the tso primary/leader has changed, try to re-campaign a primary/leader")
+		}
+
+		s.campaignLeader()
+	}
+}
+
+func (s *Server) campaignLeader() {
```

Comment:
After a discussion with @binshi-bing, later we may need to implement a TSO allocator manager to manage all keyspace groups' TSO allocators and campaign primary for each of them. For now, it's ok for me to just extract a basic single-leader election-based TSO service in this PR.

Reply:
+1